### DISCUSSION TITLE
PDFLoader

### DISCUSSION THREAD

**Discussion by chrishay** [Abraxas-365/langchain-rust#122] [General]
Hi,
Sorry for the stupid question, but once a PDF is loaded how is it used in the chain as part of the context?
Thanks

**Comment by Abraxas-365** [comment]
Hi you can check this example 

https://github.com/Abraxas-365/langchain-rust/blob/main/examples/qa_chain.rs

Also if you want to create a custom chain you can follow the same steps as the qa_chain

**Reply by Abraxas-365** [reply]
If the pdf is big, you should save the pdf on a vector store, and you can use the https://github.com/Abraxas-365/langchain-rust/blob/main/examples/conversational_retriever_chain_with_vector_store.rs

**Reply by Abraxas-365** [reply]
This is hoy you can house the vector store by itself https://github.com/Abraxas-365/langchain-rust/blob/main/examples/vector_store_postgres.rs

**Reply by Abraxas-365** [reply]
With this you can create you own custom qa retriever chain

**Reply by Abraxas-365** [reply]
@chrishay let me know if that work for you , if you want I can give you a pseudo code of how to use them in combination to create you own retriever qa chain

**Reply by chrishay** [reply]
Thanks so much for the fast response, and apologies for the stupid questions as i'm also learning rust, but is there an example where the below is using a PDF loader instead?

```
.documents(&[
            Document::new(format!(
                "\nQuestion: {}\nAnswer: {}\n",
                "Which is the favorite text editor of luis", "Nvim"
            )),
            Document::new(format!(
                "\nQuestion: {}\nAnswer: {}\n",
                "How old is Luis", "24"
            )),
        ])
```

**Reply by chrishay** [reply]
I have this example of loading the PDF which is working, I just don't know how to add it to the LLM query. Thanks again!


```
let path = "./test.pdf";

let loader = LoPdfLoader::from_path(path).expect("Failed to create PdfLoader");

    let docs = loader
        .load()
        .await
        .unwrap()
        .map(|d| d.unwrap())
        .collect::<Vec<_>>()
```

**Reply by Abraxas-365** [reply]
@chrishay If you pdf is big, i recommend you use some vector store to store the documents, and then using a qa retriver chain.

Lets say you pdf is a small one, you can do this
```rust
let path = "./test.pdf";

let loader = LoPdfLoader::from_path(path).expect("Failed to create PdfLoader");

    let your_docs = loader
        .load()
        .await
        .unwrap()
        .map(|d| d.unwrap())
        .collect::<Vec<_>>()
        .await;
        
         let llm = OpenAI::default();
    let chain = StuffDocument::load_stuff_qa(llm);
    let input = chain
        .qa_prompt_builder()
        //You could also get the documents form a retriver
        .documents(&your_docs)
        .question("How old is luis and whats his favorite text editor")
        .build();
        
       let ouput = chain.invoke(input).await.unwrap(); 
```

**Reply by Abraxas-365** [reply]
@chrishay if you want you can join the discord, and we can go to the voice channel 😄

**Reply by chrishay** [reply]
Thanks i will try and join as am getting a compile error, for .documents(&your_docs) 
 expected `&[Document]`, found `&Collect<Map<..., ...>, ...>`

**Reply by Abraxas-365** [reply]
Thatas weird, are you using feature_utils? lets jumpt to the discord

**Comment by chrishay** [comment]
> @chrishay if you want you can join the discord, and we can go to the voice channel 😄
I'm in Discord General channel Thanks :)

**Comment by Abraxas-365** [comment]
It was missing the las await to joing the streams.

```rust
    let path = "./chris.pdf";
    let loader = LoPdfLoader::from_path(path).expect("Failed to create PdfLoader");

    let your_docs = loader
        .load()
        .await
        .unwrap()
        .map(|d| d.unwrap())
        .collect::<Vec<_>>()
        .await; //this was missing

```

**Comment by chrishay** [comment]
thank you for all the help

**Reply by Abraxas-365** [reply]
No worries. Please, @chrishay  check your Discord DM.
